test/asynchronous/test_database.py

```diff
@@ -0,0 +1,778 @@
+
+        # Should send nameOnly (except on 2.6).
+        for filter in (None, {}, {"name": {"$in": ["capped", "non_capped"]}}):
+            listener.reset()
+            names = await db.list_collection_names(filter=filter)
+            self.assertIn("capped", names)
+            self.assertIn("non_capped", names)
+            command = listener.started_events[0].command
+            self.assertIn("nameOnly", command)
+            self.assertTrue(command["nameOnly"])
+
+    async def test_check_exists(self):
+        listener = OvertCommandListener()
+        client = await async_rs_or_single_client(event_listeners=[listener])
+        self.addAsyncCleanup(client.close)
+        db = client[self.db.name]
+        await db.drop_collection("unique")
+        await db.create_collection("unique", check_exists=True)
+        self.assertIn("listCollections", listener.started_command_names())
+        listener.reset()
+        await db.drop_collection("unique")
+        await db.create_collection("unique", check_exists=False)
+        self.assertTrue(len(listener.started_events) > 0)
+        self.assertNotIn("listCollections", listener.started_command_names())
+
+    async def test_list_collections(self):
+        await self.client.drop_database("pymongo_test")
+        db = AsyncDatabase(self.client, "pymongo_test")
+        await db.test.insert_one({"dummy": "object"})
+        await db.test.mike.insert_one({"dummy": "object"})
+
+        results = await (await db.list_collections()).to_list()
+        colls = [result["name"] for result in results]
+
+        # All the collections present.
+        self.assertTrue("test" in colls)
+        self.assertTrue("test.mike" in colls)
+
+        # No collection containing a '$'.
+        for coll in colls:
+            self.assertTrue("$" not in coll)
+
+        # Duplicate check.
+        coll_cnt: dict = {}
+        for coll in colls:
+            try:
+                # Found duplicate.
+                coll_cnt[coll] += 1
+                self.assertTrue(False)
+            except KeyError:
+                coll_cnt[coll] = 1
+        coll_cnt: dict = {}
+
+        # Checking if is there any collection which don't exists.
+        if (
+            len(set(colls) - {"test", "test.mike"}) == 0
+            or len(set(colls) - {"test", "test.mike", "system.indexes"}) == 0
+        ):
+            self.assertTrue(True)
+        else:
+            self.assertTrue(False)
+
+        colls = await (await db.list_collections(filter={"name": {"$regex": "^test$"}})).to_list()
+        self.assertEqual(1, len(list(colls)))
+
+        colls = await (
+            await db.list_collections(filter={"name": {"$regex": "^test.mike$"}})
+        ).to_list()
+        self.assertEqual(1, len(list(colls)))
+
+        await db.drop_collection("test")
+
+        await db.create_collection("test", capped=True, size=4096)
+        results = await (await db.list_collections(filter={"options.capped": True})).to_list()
```

Comment:
`AsyncCursors` are asynchronous iterators, so you can use `[async for result in results]` instead of calling `to_list()` here.